base/acme/src/main/java/org/dogtagpki/acme/backend/ACMEBackend.java

```diff
@@ -30,10 +34,54 @@ public void init() throws Exception {
     public void close() throws Exception {
     }
 
-    public BigInteger issueCertificate(String csr) throws Exception {
+    /**
+     * This method generates a unique ID for a certificate.
+     *
+     * By default this method will return the base64-encoded serial number
+     * of the certificate. This method may be overridden to generate a backend-
+     * specific unique ID for the certificate.
+     *
+     * @param certData Certificate data in byte array.
+     * @return Unique ID for the certificate.
+     * @throws Exception
+     */
+    public String getCertificateID(byte[] certData) throws Exception {
+
+        try (ByteArrayInputStream is = new ByteArrayInputStream(certData)) {
+
+            CertificateFactory cf = CertificateFactory.getInstance("X.509");
+            X509Certificate cert = (X509Certificate) cf.generateCertificate(is);
+            BigInteger serialNumber = cert.getSerialNumber();
+
+            return Base64.encodeBase64URLSafeString(serialNumber.toByteArray());
+        }
+    }
+
+    /**
+     * This method generates a certificate using the provided certificate signing request,
+     * then returns the certificate data in byte array.
+     *
+     * @param csr Certificate signing request.
+     * @return Certificate data in byte array.
+     * @throws Exception
+     */
+    public byte[] generateCertificate(String csr) throws Exception {
         return null;
```

Comment:
Should this raise something like `NotImplementedException` ?

Reply:
And... why `byte[]`? Should we make it return a `X509Certificate` instead, and then `getCertificateID` can take a `X509Certificate` and have an overload to take a `byte[]` and do the conversion? Just a thought.

Reply:
Yes, we should be able to throw `NotImplementedException`. Let me see if we can use `X509Certificate`.

Reply:
Fixed.